### DISCUSSION TITLE
SimpleNamespace and Protocol

### DISCUSSION THREAD

**Discussion by seasonsteven** [microsoft/pyright#9546] [General]
Consider this sample:

```python
from typing import NamedTuple, Protocol
from types import SimpleNamespace


class ISwitch(Protocol):
    @property  # Intention: read required, write optional
    def off(self) -> str: ...

    @property
    def on(self) -> str: ...


class SwitchNT(NamedTuple):
    off: str
    on: str


class Switch:
    def __init__(self, off: str, on: str) -> None:
        self.off = off
        self.on = on


switch1: ISwitch = SwitchNT(off="A", on="B")
switch2: ISwitch = Switch(off="A", on="B")
switch3: ISwitch = SimpleNamespace(off="A", on="B")

switches1: list[ISwitch] = [SwitchNT(off="A", on="B")]
switches2: list[ISwitch] = [Switch(off="A", on="B")]
switches3: list[ISwitch] = [SimpleNamespace(off="A", on="B")]
```

The above passes mypy but pyright does not accept the SimpleNamespace instance.

```
Type "SimpleNamespace" is not assignable to declared type "ISwitch"
  "SimpleNamespace" is incompatible with protocol "ISwitch"
    "off" is not present
    "on" is not present  (reportAssignmentType)
Type "list[SimpleNamespace]" is not assignable to declared type "list[ISwitch]"
  "SimpleNamespace" is incompatible with protocol "ISwitch"
    "off" is not present
    "on" is not present  (reportAssignmentType)
 ```
   
Is it reasonable to expect pyright to consider them compatible?

**Comment by erictraut** [comment]
> Is it reasonable to expect pyright to consider them compatible?

No, a `SimpleNamespace` class doesn't have any declared any attributes or types, so it is not compatible with a protocol that requires certain attributes with certain types.

If you want to write type-safe code, I recommend avoiding `SimpleNamespace`. Use a `NamedTuple` or `dataclass` instead.

**Comment by brianschubert** [comment]
For context, the reason mypy accepts this is because `SimpleNamespace` has a `__getattribute__` definition in typeshed, which causes mypy to consider all possible attributes to exist and have type `Any` in subtype checks:

https://github.com/python/typeshed/blob/d70fad09dd7012bd43304e1c67b7599558b90c39/stdlib/types.pyi#L325
